python/parquet_tools/common/schema.py

```diff
@@ -0,0 +1,700 @@
+        #
+        for cname, sqltype, nptype in pqcols:
+            n += 1
+            if cname in Schema._colNotNull:
+                nntext = ' NOT NULL'
+            else:
+                nntext = ''
+            if not sqltype or sqltype == nptype.upper():
+                print(fmt1.format(n, cname, nptype + nntext))
+            else:
+                print(fmt2.format(n, cname, nptype, sqltype + nntext))
+
+    # *************************************************************************
+    # *                        _ g e t _ d e c t y p e                        *
+    # *************************************************************************
+    @classmethod
+    def _get_dectype(cls, dargs, cname):
+        """Convert decimal type to appropriate numpy type.
+
+        Parameters
+        ----------
+        dargs : 'string'
+            The characters after 'decimal(' which should be 'p[,s])'. Where
+            'p' is the precision and 's' is the scale.
+        cname : 'string'
+            The associated column name.
+
+        Returns
+        -------
+        result : 'string'
+            The returned string is the generic numpy datatype to be used to
+            represent the decimal type. A null string indicates that the
+            appropriate type could not be determined.
+
+        Notes
+        -----
+        - This method does not check whether or not the type is fully
+          representable in the returned type as these will be caught during
+          actual conversion.
+        - When s is missing or is 0, and integer type is returned whose
+          size is dependent on the 'p', as follows:
+           0 <= p <  3: int8
+           3 <= p <  5: int16
+           5 <= p < 10: int32
+          10 <= p:      int64 (note max p is 18 but we don't check)
+        - When 's' is specified, then we return float32 when p < 7 and
+          float64 otherwise. This is an imperfect data type for decimals
+          as it is not very precise but it's the best one we have.
+        """
+
+        # Find the closing paren a strip off the remaining text
+        #
+        n = dargs.find(')')
+        if n < 1:
+            return ''
+        xargs = dargs[0:n]
+
+        # Split this into one or two args and convert
+        #
+        pspec = xargs.split(',')
+        try:
+            pval1 = int(pspec[0])
+        except Exception:  # We don't care what the exception is
+            return ''
+        if pval1 < 0:
+            return ''
+        if len(pspec) == 1:
+            pval2 = 0
+        else:
+            try:
+                pval2 = int(pspec[1])
+            except Exception:  # We don't care what the exception is
+                return ''
+
+        # Do check for valid scale value
+        #
+        if pval2 < 0:
+            return ''
+
+        # Return the an int type if this has no digits after the decimal point
+        #
+        if pval2 == 0:
+            if pval1 < 3:
```

Comment:
this is all kind of magic-numbery

Reply:
Unfortunately, it sort of has to be since you are trying to compute the integer type based on a decimal type of specific precision. The comments explain how that works and the code merely is a literal implementation of the comments.